lib/trino-plugin-toolkit/src/main/java/io/trino/plugin/base/security/FileBasedSystemAccessControl.java

```diff
@@ -953,7 +965,7 @@ public void checkCanExecuteFunction(SystemSecurityContext systemSecurityContext,
     @Override
     public void checkCanExecuteFunction(SystemSecurityContext systemSecurityContext, FunctionKind functionKind, CatalogSchemaRoutineName functionName)
     {
-        if (functionKind == TABLE) {
+        if (functionKind == TABLE && !checkTableFunctionPermission(systemSecurityContext, functionName)) {
```

Comment:
May be that's good enough for table functions, but IIRC the eventual goal is to use `CatalogSchemaRoutineName` for `functionName` for other `FunctionKind`s too.

The proposed config syntax is too specific to table functions too:
```
{
    "table_functions": [
        {
            "user": "admin",
            "catalog": "postgresql",
            "schema": "system",
            "table_function": "query"
        }
    ]
}
```
I think it worth to go with something like 
```
{
    "function": [
        {
            "user": "admin",
            "function_kind": "TABLE",
            "catalog": "postgresql",
            "schema": "system",
            "function": "query"
        }
    ]
}
```

I'd base code for functions on the code for tables, e.g.
```
    @Override
    public void checkCanExecuteFunction(SystemSecurityContext systemSecurityContext, FunctionKind functionKind, CatalogSchemaRoutineName function)
    {
        // TODO config to disable for old function kinds
        AccessMode requiredCatalogAccess = FunctionKind.TABLE.equals(functionKind) ? ALL : READ_ONLY;
        if (!checkFunctionPermission(systemSecurityContext, function, requiredCatalogAccess, FunctionPrivilege.SELECT)) {
            denyExecuteFunction(function.toString());
        }
    }

    private boolean checkFunctionPermission(SystemSecurityContext context, CatalogSchemaRoutineName function, AccessMode requiredCatalogAccess, FunctionPrivilege checkPrivilege)
    {

        if (!canAccessCatalog(context, function.getCatalogName(), requiredCatalogAccess)) {
            return false;
        }

        if (INFORMATION_SCHEMA_NAME.equals(function.getSchemaName())) {
            return true;
        }

        Identity identity = context.getIdentity();
        for (CatalogFunctionAccessControlRule rule : functionRules) {
            if (rule.matches(identity.getUser(), identity.getEnabledRoles(), identity.getGroups(), function)) {
                return rule.getPrivileges().contains(checkPrivilege);
            }
        }
        return false;
    }
```

Reply:
I think it's a valid concern. If the plan is to have other catalog/schema function kinds in the future then the current syntax would require backward incompatible change or keeping `TABLE` as the default function kind. Neither of which is desired.

Reply:
@lukasz-walkiewicz I confirm that the discussed method `checkCanExecuteFunction()` will be used for all function kinds in the future.

@ssheikin, regarding this part:
```
AccessMode requiredCatalogAccess = FunctionKind.TABLE.equals(functionKind) ? ALL : READ_ONLY;
        if (!checkFunctionPermission(systemSecurityContext, function, requiredCatalogAccess, FunctionPrivilege.SELECT)) {
            denyExecuteFunction(function.toString());
        }
```
requiring catalog access `ALL` is safe, but might be unnecessarily restrictive for some Table Functions (think: sequence generator, random generator etc. which do not access any data even for reading).

I also wonder how it fits the future model where all built-in functions are provided by the Global catalog. The built-in table functions should be easily available to anyone, without granting them that level of access to the catalog.

Ideally, the required catalog access should be set for each table function independently.

What does `FunctionPrivilege.SELECT` refer to?

Reply:
Ugly initial draft based on file based access control of tables: https://github.com/trinodb/trino/pull/13903 . NOT to review. Only for sake of history.

Reply:
@dain your suggestions would be very helpful here:
1. What catalog access should we require for table functions? How about the built-in functions provided by the Global catalog?
2. Could we possibly specify the required access per table function? Some functions read data (query pass-through), some might modify the catalog, and some don't touch anything (sequence).

Reply:
@huberty89 Why?

Reply:
Hmm, we could support all function kinds but in that case when user defined at least one rule about `TABLE` function it's required to open access to the rest of functions by adding the rule:
```
{
    "functions": [
        {
            "function_kinds": [
                "SCALAR",
                "AGGREGATE",
                "WINDOW"
            ],
            "privileges": [
                "EXECUTE"
            ]
        }
    ]
}
```

Reply:
I think it's ok, but we have to have a possibility to return to well known behaviour(and, actually, such behaviour should be disabled by default for backwards compatibility reasons).
https://github.com/trinodb/trino/pull/13713#discussion_r960653515

Reply:
I'm not sure if additional property which will change the behavior of entered rules is a good idea - this json format is already quite complicated.

Reply:
I add support for every function kind

Reply:
@kasiafi 
>Ideally, the required catalog access should be set for each table function independently.

Now I think this makes sense in case of UDFs as well.
I will take a look into it.

Reply:
In theory user might prepare UDF which will make some modifications - this is such edge case probably we should not take care of this.

Reply:
I guess it's time for that `eventually` https://github.com/trinodb/trino/pull/12544#discussion_r906139814

Reply:
or at least

> this is such edge case probably we should not take care of this

.... YET